### PR TITLE
Scenario-set construction from traces: Clio facets -> cluster -> select -> synthesize -> verify

### PR DESCRIPTION
**Pull request by kfallah**
## What

Adds representative eval-scenario-set construction from telemetry traces — the harness can now decide *which* scenarios to run, not just execute them. Design A from the research survey ([.agents/docs/research/scenario-set-construction-sota.md](../blob/feature/scenario-set-construction/.agents/docs/research/scenario-set-construction-sota.md)): Clio-style facets → cluster → hybrid-allocation medoid selection → WildBench-style synthesis → Trajectory2Task-style closed-loop verification.

### `wmh/scenarios/` (new package)
- **facets.py** — LLM facet extraction per trace (task summary ≤30 words, outcome, failure category) + deterministic tool-call signature; facet embeddings, never raw-trace embeddings (intent washout).
- **clustering.py** — numpy k-means (kmeans++, cosine, deterministic) + LLM cluster naming.
- **selection.py** — SemDeDup within clusters, hybrid budget allocation (70% traffic-proportional / 30% coverage with uncovered-cluster priority), medoid + farthest-first picks, failure-category pinning, corpus-mass weights summing to 1.
- **synthesis.py** — trace → `EvalScenario` (self-contained task, seed `EnvState`, judgeable checklist, provenance); `ScenarioSet` JSON artifact.
- **verification.py** — `ChecklistJudge`, back-agreement (judge grades the source trajectory; must match recorded outcome) + solvability (baseline-agent rollout in the world model).

### Supporting
- `wmh/env/llm_agent.py` — reusable baseline LLM agent (JSON tool-call per turn).
- `wmh/research/scenario_recovery.py` + `scenario_fidelity.py` — purity/ARI vs ground truth; score-matrix + MAE/Spearman/Kendall vs random-K baselines (all metrics mean of 3 passes).
- CLI: `wmh scenarios build` / `wmh scenarios verify [--drop]`.
- Bedrock provider: Amazon Nova request/response support (Claude-on-Bedrock is gated on this account; e2e ran on Nova).

## E2E verification (tau-bench, small models)

60 traces (20/domain), Nova Lite pipeline + world model, Nova Micro/Lite × t0/t0.9 agents, Titan v2 embeddings. Full writeup with limitations: [.agents/docs/research/scenario-set-e2e-tau-bench.md](../blob/feature/scenario-set-construction/.agents/docs/research/scenario-set-e2e-tau-bench.md).

- **Predictive fidelity** (K=8 vs 30-pool, 360 episodes): ours MAE **0.035** vs random-K mean 0.048 (beats 2/3 seeds). Rank correlations underpowered — the 4 agent configs span only 0.057 true score spread.
- **Closed-loop verification**: 62% solvable, 62% back-agreement — the `--drop` filter works as intended.
- **Domain recovery**: raw-digest embeddings win (1.0 vs 0.667 purity) by construction (tool schemas encode domain); facets capture the intent axis. Real task-identity recovery needs a multi-rollout corpus — follow-up.

## Known gaps / follow-ups
- Corpus has one rollout per task → task-identity recovery untestable here; TRAIL or a multi-rollout capture is the follow-up.
- Rank metrics need a wider agent capability spread.
- Selection layer is a clean seam for facility-location (Design B) and IRT compression (Design C).

Gates: `ruff` clean, `ty` clean, 392 passed / 7 skipped. `/ready-for-merge` not yet run.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: wmh/scenarios/verification.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    source = next((traces_by_id[tid] for tid in scenario.provenance if tid in traces_by_id), None)
+    if source is not None and scenario.checklist:
+        source_result = judge.score(scenario.task, scenario.checklist, source.steps)
+        judge_success_on_source = source_result.success
+        recorded = _recorded_outcome(source, scenario.source_outcome)
+        if recorded is not Outcome.UNKNOWN:
+            back_agreement = source_result.success == (recorded is Outcome.SUCCESS)
+
+    # Solvability: one baseline-agent rollout in the world model, graded by the same judge.
+    env = WorldModelEnv(world_model)
```
[the remark this answers is not in the document]
Comment (kfallah) [reply wmh/scenarios/verification.py:200]: Fixed in c11dd0a: added `WorldModel.frozen()` (suspends online index enrichment); `verify_scenarios` runs entirely inside it, so rollout steps are never indexed and verdicts are order-independent. Regression test asserts `retriever.add` is never called during verification and that serving enrichment resumes after the block.

File: wmh/research/scenario_fidelity.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    workers: int = 4,
+) -> ScoreMatrix:
+    """Roll every agent over every pool scenario `passes` times; cell = mean checklist pass rate.
+
+    Rollouts run on a small thread pool (provider calls are I/O bound); each rollout opens its own
+    world-model session so episodes never share state.
+    """
+
+    def one_rollout(agent: Agent, scenario: EvalScenario) -> float:
+        env = WorldModelEnv(world_model)
```
[the remark this answers is not in the document]
Comment (kfallah) [reply wmh/research/scenario_fidelity.py:74]: Fixed in c11dd0a: `score_matrix` wraps the whole thread-pooled matrix in `world_model.frozen()` — no cell's generated steps can become another cell's retrieved demos. New test runs a 2-agent × 2-scenario × 2-pass matrix and asserts zero retriever enrichment.

File: wmh/cli/app.py [hunk 3]
```diff
@@ -771,6 +786,162 @@ def _eval_report_payload(report: EvalReport) -> dict[str, object]:
...
+            f"{verdict.rollout_pass_rate:.2f}",
+        )
+    _console.print(table)
+    _console.print(
+        f"back-agreement {report.back_agreement_rate:.0%}, solvable {report.solvable_rate:.0%} "
+        f"over {len(report.verdicts)} scenarios"
+    )
+    if drop:
+        verified = {v.scenario_id for v in report.verdicts if v.ok}
+        scenario_set.scenarios = [s for s in scenario_set.scenarios if s.scenario_id in verified]
```
[the remark this answers is not in the document]
Comment (kfallah) [reply wmh/cli/app.py]: Fixed in c11dd0a: `--drop` now goes through `ScenarioSet.retain()`, which renormalizes surviving weights to sum to 1 and zeroes `corpus_coverage`/`coverage_tau` (coverage can't be recomputed at verify time — the facet embeddings are gone — so stale numbers are invalidated rather than kept). Covered by tests incl. the drop-everything edge case.

File: wmh/scenarios/synthesis.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        if parsed is not None and parsed.task.strip():
+            task = parsed.task.strip()
+            seed_state = EnvState(scratchpad=parsed.initial_state.strip())
+            checklist = [item.strip() for item in parsed.checklist if item.strip()]
+        else:
+            task = facet.task_summary
+            seed_state = EnvState()
+            checklist = []
+        return EvalScenario(
+            scenario_id=f"scenario-{trace.trace_id[:12]}",
```
[the remark this answers is not in the document]
Comment (kfallah) [reply wmh/scenarios/synthesis.py]: Fixed in c11dd0a: scenario_id now uses the full trace_id.

File: wmh/cli/app.py [hunk 5]
```diff
@@ -771,6 +786,162 @@ def _eval_report_payload(report: EvalReport) -> dict[str, object]:
...
+    out: str = typer.Option("scenarios.json", "--out", help="Where to write the scenario set."),
+    budget: int = typer.Option(20, help="Number of scenarios to construct."),
+    k: int = typer.Option(None, help="Cluster count (default: sqrt(corpus size))."),
+    limit: int = typer.Option(None, help="Only use the first N ingested traces (cost control)."),
+    provider: str = typer.Option("bedrock", "--provider", help="LLM for facets/naming/synthesis."),
+    model: str = typer.Option("us.anthropic.claude-opus-4-8", help="Model id."),
+    region: str = typer.Option(None, help="AWS region (Bedrock)."),
+    embed_provider: str = typer.Option(
+        "hashing", help="Facet embedder: hashing (offline) | bedrock | openai | azure_openai."
+    ),
```
[the remark this answers is not in the document]
Comment (kfallah) [reply wmh/cli/app.py:980]: Fixed in c11dd0a: help text now states the limitation. Worth noting HashingEmbedder is a hashed character-trigram bag (lexical), not random projection — but the substance of the comment stands: it clusters by wording, not meaning, and the help text now says to prefer a semantic embedder for real corpora. Kept as default for parity with `wmh build` (offline zero-config philosophy).

File: wmh/scenarios/clustering.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    matrix = normalize_rows(embeddings)
+    n = matrix.shape[0]
+    if k < 1:
+        raise ValueError(f"k must be >= 1, got {k}")
+    if k >= n:
+        return np.arange(n, dtype=np.int64)
+    rng = np.random.default_rng(seed)
+    centroids = _kmeans_pp_init(matrix, k, rng)
+    labels = np.zeros(n, dtype=np.int64)
+    for _ in range(_KMEANS_ITERS):
```
[the remark this answers is not in the document]
Comment (kfallah) [reply wmh/scenarios/clustering.py]: Fixed in c11dd0a: sentinel is now `np.full(n, -1)`.

**Comment by kfallah** [comment]
**Correction to the Test-2 claims in the PR description** (analysis in `.agents/docs/research/selection_vs_random_correction.json`, cb21500):

The 3-seed random baseline was underpowered and the e2e script had a selection bug (re-clustered the pool at default k≈5 instead of the 8 build clusters — `build_scenario_set` itself is correct). Recomputed offline against 2000 random draws on the saved score matrix:

| selection | pool-mean MAE | percentile vs random | clusters covered |
|---|---|---|---|
| uniform random K=8 | 0.045 mean | — | ~5.2/8 |
| ours as run (buggy) | 0.035 | 61st pct — random-equivalent | 4/8 |
| ours corrected | 0.100 | 3rd pct — **worse than random** | **7/8** |
| stratified-random control (same allocation, random within cluster) | 0.054 | ≈ random | 7/8 |

Reading: for estimating the pool-mean score, intent-space **medoids are score-biased** (the stratified-random control shows the allocation itself is fine). The selection's demonstrated value is **coverage** (7/8 intent clusters vs ~5/8 random; an untested intent is an undetectable regression) **and auditability**, not mean-score calibration. Follow-up: add an unbiased `pick="random"` mode to `hybrid_select` for the estimation use-case; Design-C IRT remains the right tool for score-predictive compression.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by kfallah** [comment]
**Inverse-distillation verification complete** (charter: `.agents/docs/proposals/inverse-distillation-charter.md`, journal in claas-verl `experiments/tau/07_03_2026_wm-mined-scenario-distill.md`):

Traces → mined+verified scenario pools (65 train / 21 eval, disjoint by source trace) → privileged teacher in the frozen WM (56 episodes, 276 turn samples, checklist rejection) → LoRA SFT (Qwen3.5-9B) → paired eval.

| arm | WM success | WM pass-rate | real-env success | real pass-rate | real tool-err |
|---|---|---|---|---|---|
| base | 31.7% | 0.537 | 3.2% | 0.127 | 66% |
| **wmh-sft** | **46.0%** | **0.631** | 4.8% | **0.292** | **53%** |

+14.3pts on held-out mined evals (11W/5L/5T paired; retail 0→33%, telecom 37→48%). Real-env binary success flat (strict schemas + synthesized-entity mismatch) but soft metrics transfer (pass-rate 2.3×, −13pts invalid calls, 0-step episodes 9→0).

Two judge bugs found by transcript audits and fixed in this branch with regression tests: final-message visibility and reasoning-judge verdict truncation.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by kfallah** [comment]
**Scope clarification on the distillation result above** (after review discussion): the +14.3pts validates the *plumbing* — mined scenarios are executable, gradeable, and trainable-on, and the held-out yardstick + rejection filter are mining products — but it does **not** demonstrate that representative mining beats random task selection for training, and the teacher's reference-trace hint is a confound (hinted demos are partly cleaned re-renderings of the source traces). The PR's reviewable claim is the mining layer itself (`wmh/scenarios/`); the experiment in `.agents/` is evidence of end-to-end usability. Decisive ablations (random-scenario arm, unhinted-teacher arm; ~$15 each) are listed as pending in the journal.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: wmh/scenarios/verification.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kfallah) [reply wmh/scenarios/verification.py:200]: Addressed: `verify_scenarios` wraps all rollouts in `world_model.frozen()` (wmh/scenarios/verification.py:171), so generated steps never enter the shared retriever and later scenarios cannot retrieve earlier rollouts.

File: wmh/research/scenario_fidelity.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kfallah) [reply wmh/research/scenario_fidelity.py:74]: Addressed: the score matrix runs under `world_model.frozen()` with a per-episode `WorldModelEnv` session (wmh/research/scenario_fidelity.py:86), so cells cannot leak state into each other through the retriever.

File: wmh/cli/app.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kfallah) [reply wmh/cli/app.py]: Addressed: `ScenarioSet.retain` now renormalizes survivor weights to sum to 1 and zeroes the stale `corpus_coverage` (rationale in its docstring, wmh/scenarios/synthesis.py:51); `--drop` saves the set only after `retain`.

**Comment by kfallah** [comment]
## Filtered-BC ablation: mined vs random scenario selection — result

The decisive experiment for whether mined scenarios beat randomly-selected scenarios as a *training* signal. Clean ablation: same synthesis + validation for both pools (only SELECTION differs), same self-BC collection (base Qwen3.5-9B, 6 samples/scenario, keep ≤2 judge-passing episodes), identical LoRA hyperparams, same eval (21 held-out scenarios × 3 passes, gpt-5.4 WM, Opus 4.8 judge).

| arm | success | pass-rate | retail | telecom | train data |
|---|---|---|---|---|---|
| base | 27.0% | 0.425 | 22.2% / 0.33 | 27.8% / 0.44 | — |
| bc-mined | 22.2% | 0.488 | 0.0% / 0.46 | 25.9% / 0.49 | 73 eps / 321 samples |
| bc-random | **36.5%** | **0.581** | 22.2% / 0.54 | 38.9% / 0.59 | 48 eps / 206 samples |

Paired per-scenario (n=21, two-sided sign tests): mined vs random success **3W/8L/10T** (Δ −14.3 pts, p=0.23), pass-rate **7W/11L/3T** (Δ −0.093, p=0.48). Random vs base: +9.5 pts success, +0.156 pass-rate (p=0.39 / p=0.10).

**Verdict: the mining-helps-training hypothesis is not supported.** No comparison reaches significance at n=21, and every point estimate favors random — despite the mined arm having *more* training data (mined scenarios are more collectible under self-BC: 73 vs 48 kept episodes) and the eval pool itself being mined, which if anything should bias toward the mined arm.

What this does and doesn't say:
- It does NOT undercut the mining pipeline's other role — validated eval-set construction — nor the earlier end-to-end result (which used a privileged teacher, a different treatment).
- It DOES say cluster-based mined selection is not a better *training-data selector* than uniform random in this self-BC setup. Post-hoc analysis points at two data-backed mechanisms: (1) **distribution shift** — diversity-seeking selection flattened the domain mix (mined kept-episodes: 47% telecom, 15% airline-which-the-eval-lacks) while random matched the telecom-heavy corpus/eval (87% telecom); (2) **within-domain concentration** — the gap persists within telecom alone (paired 3W/7L/8T, −13 pts) where both arms had comparable data, and 13 of 35 mined telecom scenarios are MMS-troubleshooting variants. Cluster-diverse ≠ instance-diverse. A corpus-proportional-quota variant of mining would isolate (1) from (2).
- Statistical honesty: n=21 scenarios → wide CIs; the claim is "no advantage, direction reversed", not "random is significantly better".

**Update (2026-07-06b): random-seed variance test.** The single random arm above turned out to be a lucky draw. Two more full random arms (independent seeds, identical pipeline) landed at 27.0% success each — equal to base. Random over 3 seeds: **30.2% ± 5.5 success, 0.564 ± .035 pass-rate**. bc-mined-v2 (below) sits at **exactly the random mean** on success (30.2%; paired vs per-scenario mean-of-3-seeds: 6W/6L/9T, Δ +0.000) and −0.027 behind on pass-rate. Sharpened verdict: summary-only mined selection is genuinely worse than random (−8 pts vs the mean); capability-enriched mined selection is at parity; nothing beats random yet; and self-BC itself is worth about +3 pts success / +0.14 pass-rate on average, with ±5.5-pt draw-to-draw noise that any future improvement claim must clear.

**Update (2026-07-06a): enriched-embedding rematch.** We re-ran the mined arm with one fix — facet embeddings now carry domain + tool signature instead of the bare task summary (`454d93f`), so clustering groups by capability rather than phrasing. Result: **bc-mined-v2 = 30.2% success / 0.537 pass-rate** — recovers most of the deficit (paired vs random now 3W/5L/13T, −6.3 pts, vs 3W/8L/10T, −14.3 pts before) and outright fixes the retail collapse (0% → 33.3%, best of all four arms). Mined selection is now statistically indistinguishable from random, but still doesn't beat it; the residual gap is entirely telecom, i.e. the still-unfixed 70/30 allocator drift. Next single-variable test: corpus-proportional domain quotas in selection.

Artifacts in `.agents/docs/research/distill/` @ fd6ed5b (pools incl. v2, kept episodes, per-arm eval JSONs). Full protocol + audit trail in the experiment journal.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
